src/main/java/org/jboss/pnc/common/http/PNCHttpClient.java

```diff
@@ -0,0 +1,181 @@
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.time.Duration;
+import java.time.temporal.ChronoUnit;
+import java.util.function.Supplier;
+
+import static org.jboss.pnc.api.constants.HttpHeaders.AUTHORIZATION_STRING;
+
+/**
+ * Wrapper around {@link HttpClient} that can handle PNC {@link Request}s.
+ */
+@Slf4j
+public class PNCHttpClient {
+    private final HttpClient client;
+    private final ObjectMapper objectMapper;
+    private RetryPolicy<HttpResponse<String>> retryPolicy;
+    @Setter
+    private Supplier<String> tokenSupplier;
+
+    public PNCHttpClient(RetryConfig retryConfig) {
+        this(new ObjectMapper(), retryConfig);
+    }
+
+    public PNCHttpClient(ObjectMapper objectMapper, RetryConfig retryConfig) {
+        this.client = HttpClient.newBuilder().followRedirects(HttpClient.Redirect.NORMAL).build();
+        this.objectMapper = objectMapper;
+
+        retryPolicy = RetryPolicy.<HttpResponse<String>> builder()
+                .withBackoff(retryConfig.backoffInitialDelay(), retryConfig.backoffMaxDelay(), ChronoUnit.SECONDS)
+                .withMaxRetries(retryConfig.maxRetries())
+                .withMaxDuration(Duration.of(retryConfig.maxDuration(), ChronoUnit.SECONDS))
+                .onSuccess(e -> log.debug("Request successfully sent, response: {}", e.getResult().statusCode()))
+                .onRetry(
+                        e -> log.debug(
+                                "Retrying (attempt #{}), last response was.",
+                                e.getAttemptCount(),
+                                e.getLastException()))
+                .onFailure(e -> log.warn("Request couldn't be sent.", e.getException()))
+                .build();
+    }
+
+    /**
+     * Tries to send the request once, reporting any error immediately.
+     *
+     * @param request Request to be sent.
+     * @throws PNCHttpClientException When there is error sending the request.
+     */
+    public void trySendRequest(Request request) {
+        trySendRequest(request, request.getAttachment());
+    }
+
+    /**
+     * Tries to send the request once replacing the payload from {@link Request#getAttachment()} with provided payload,
+     * reporting any error immediately.
+     *
+     * @param request Request to be sent.
+     * @param payload Payload to be used instead of {@link Request#getAttachment()}
+     * @throws PNCHttpClientException When there is error sending the request.
+     */
+    public void trySendRequest(Request request, Object payload) {
+        HttpRequest httpRequest = prepareHttpRequest(request, payload);
+        doSendRequest(httpRequest);
+    }
+
+    /**
+     * Sends the request, retrying according to the retry config if there is any error.
+     *
+     * @param request Request to be sent.
+     * @throws PNCHttpClientException When there is error sending the request.
+     */
+    public void sendRequest(Request request) {
```

Comment:
Can you add _throws PNCHttpClientException_. I know that it's runtime but still a good information if you're using the method.

Reply:
it's in javadoc, but sure